capsules/src/hmac.rs

```diff
@@ -0,0 +1,503 @@
+                                buf.copy_from_slice(d.as_ref().get(0..static_buffer).unwrap());
+                            }
+                        });
+
+                        // Add the data from the static buffer to the HMAC
+                        if let Err(e) = self.hmac.add_data(self.data_buffer.take().unwrap()) {
+                            return e.0;
+                        }
+                    }
+                    None => {
+                        return ReturnCode::ERESERVE;
+                    }
+                };
+
+                ReturnCode::SUCCESS
+            })
+            .unwrap_or_else(|err| err.into())
+    }
+}
+
+impl<H: digest::Digest<'a, T> + digest::HMACSha256, T: DigestType> digest::Client<'a>
+    for HmacDriver<'a, H, T>
+{
+    fn add_data_done(&'a self, _result: Result<(), ReturnCode>, data: &'static mut [u8]) {
+        self.appid.map(move |id| {
+            self.apps
+                .enter(*id, move |app, _| {
+                    let mut data_len = 0;
+                    let mut static_buffer = 0;
+
+                    self.data_buffer.replace(data);
+
+                    self.data_buffer.map(|buf| {
+                        match app.data.as_ref() {
+                            Some(d) => {
+                                // Determine the size of the static buffer we have
+                                static_buffer = buf.len();
+                                // Determine how much data we have already copied
+                                let copied_data = self.data_copied.get();
+
+                                data_len = d.as_ref().len();
+
+                                // If we have more data then the static buffer
+                                if data_len > copied_data {
+                                    // Copy the data into the static buffer
+                                    if data_len < (copied_data + static_buffer) {
+                                        // If the static buffer is larger then the data, just copy the rest of the slice
+                                        for (i, d) in d
+                                            .as_ref()
+                                            .get(copied_data..)
+                                            .unwrap()
+                                            .iter()
+                                            .enumerate()
+                                        {
+                                            buf[i] = *d;
+                                        }
+                                    } else {
+                                        // Copy the entire buffers worth of data
+                                        buf.copy_from_slice(
+                                            d.as_ref()
+                                                .get(copied_data..(copied_data + static_buffer))
+                                                .unwrap(),
```

Comment:
Same a above regarding `get` + `unwrap`. I don't even think you need the `as_ref` part.